src/builtins/BuiltinIterator.cpp

```diff
@@ -183,11 +185,144 @@ static Value builtinIteratorMap(ExecutionState& state, Value thisValue, size_t a
     IteratorRecord* iterated = IteratorObject::getIteratorDirect(state, O.asObject());
     // Let result be CreateIteratorFromClosure(closure, "Iterator Helper", %IteratorHelperPrototype%, « [[UnderlyingIterator]] »).
     // Set result.[[UnderlyingIterator]] to iterated.
-    IteratorHelperObject* result = new IteratorHelperObject(state, iteratorMapClosure, iterated, new IteratorMapData(mapper));
+    IteratorHelperObject* result = new IteratorHelperObject(state, iteratorMapClosure, iterated, new IteratorData(mapper));
     // Return result.
     return result;
 }
 
+static std::pair<Value, bool> iteratorFilterClosure(ExecutionState& state, IteratorHelperObject* obj, void* data)
+{
+    // Let closure be a new Abstract Closure with no parameters that captures iterated and predicate and performs the following steps when called:
+    // Let counter be 0.
+    // Repeat,
+    //   Let value be ? IteratorStepValue(iterated).
+    //   If value is done, return undefined.
+    //   Let selected be Completion(Call(predicate, undefined, « value, 𝔽(counter) »)).
+    //   IfAbruptCloseIterator(selected, iterated).
+    //   If ToBoolean(selected) is true, then
+    //      Let completion be Completion(Yield(value)).
+    //      IfAbruptCloseIterator(completion, iterated).
+    //   Set counter to counter + 1.
+    IteratorData* closureData = reinterpret_cast<IteratorData*>(data);
+    IteratorRecord* iterated = obj->underlyingIterator();
+    Value predicate = closureData->callback;
+
+    while (true) {
+        auto value = IteratorObject::iteratorStepValue(state, iterated);
+        if (!value) {
+            iterated->m_done = true;
+            return std::make_pair(Value(), true);
+        }
+
+        Value argv[2] = { value.value(), Value(closureData->counter) };
+        Value selected;
+        try {
+            selected = Object::call(state, predicate, Value(), 2, argv);
+        } catch (const Value& e) {
+            IteratorObject::iteratorClose(state, iterated, e, true);
+        }
+
+        closureData->counter = StorePositiveNumberAsOddNumber(closureData->counter + 1);
+        if (selected.toBoolean()) {
+            return std::make_pair(value.value(), false);
+        }
```

Comment:
This part seems to be related to the `generator` feature, but it's not sufficient
we may update it further later on